### ISSUE TITLE
create_db.py: break up current database into multi-table DB

### ISSUE THREAD

**Issue by cmoussa1** [flux-framework/flux-accounting#1]
Currently, **create_db.py** only creates one table that holds all information returned by `flux job list-inactive`

**write_jobs.py** writes all job record information to that one table, _inactive_. We will want to convert this into a multi-table database, one that probably contains the following tables to start:

1. **job table** and/or **job step table**
- job statistics for every job (job id, user, group, project/wckey, nodes, submit time, start time, end time, elapsed time, hosts)

2. **node state changes with reasons for entering a failed state**

3. **table for tracking wckeys**

4. **a user table to differentiate different roles such as an operator, coordinator, and admin role**

**Comment by dongahn** [comment]
Thank you for creating the repo and import your first code, cmoussa1. This is a giant step!

**Comment by chu11** [comment]
Apologies, I got newbie questions :-)

> table for tracking wckeys

I assume this is the table that indicates things like "project X gets 20% time on the machine"?

> a user table to differentiate different roles such as an operator, coordinator, and admin role

I assume so certain operators/coordinators/etc. have permissions for modifying certain tables?  Add add user to project X?  Increase Project Y percent time on cluster?

> node state changes with reasons for entering a failed state

I'm not sure what this is for.  Like if 10 nodes in the cluster are down for awhile, 20% time for a group needs to be turned down to 19.68% time?  Or perhaps I'm not understanding "node state changes".

**Comment by cmoussa1** [comment]
> I assume this is the table that indicates things like "project X gets 20% time on the machine"?

Yes, I think so. My understanding is that WCKey is what users charge their jobs against. Although, I believe the resource WCKeys are associated with resources other than just time; for example, the max number of nodes they can use to run jobs.

> I assume so certain operators/coordinators/etc. have permissions for modifying certain tables? Add add user to project X? Increase Project Y percent time on cluster?

From looking at the user table I was given access to, it looks like roles are broken up into three levels of privilege:

1. operator (people who can just run jobs on the machine)
2. coordinator (I'm guessing these types of people can add a user to a group or account to run jobs)
3. admin (just like the coordinator, but can edit limits for group accounts, enforce policies on users, change priorities for jobs, etc.)

> I'm not sure what this is for. Like if 10 nodes in the cluster are down for awhile, 20% time for a group needs to be turned down to 19.68% time? Or perhaps I'm not understanding "node state changes".

Yeah, I'm not sure this was the best description for what I think they were trying to explain. In Slurm's accounting database, they have a "\<cluster name\>_event_table", which is structured like this:

| time_start | time_end | node_name | cluster_nodes | reason | reason_uid | state | tres |
| ---------- | --------- | ------------| -------------- | ------- | ---------- | ------ | ---- |
| 1570488842 |          0 |           | opal[1-2,63-183] | Cluster Registered TRES       | 4294967294 |     0 | 1=4428,2=123,3=0,4=123,5=4428,6=0,7=0,8=0 |
| 1570488842 |          0 | opal1     |                  | Node unexpectedly rebooted    |        101 |     1 | 1=36,2=1,5=36                             |
| 1578369333 | 1578369777 | opal100   |                  | Not responding                |        101 |  2049 | 1=36,2=1,5=36                             |

Basically, when a node goes down or something breaks, an event gets written to this table explaining what happened; I'm guessing so that these nodes don't try to get scheduled when resources are requested. 

I'm pretty unfamiliar with some of the data in these tables myself, so that's why I was planning on asking more questions like this next week when I meet with Ryan 😃.

**Comment by chu11** [comment]
> Basically, when a node goes down or something breaks, an event gets written to this table explaining what happened; I'm guessing so that these nodes don't try to get scheduled when resources are requested.

Ahh, its possible this is what is used for `sinfo` and its "Reason" output.  Probably not relevant to bank/accounting is my assumption.

**Comment by grondo** [comment]
> Ahh, its possible this is what is used for sinfo and its "Reason" output. Probably not relevant to bank/accounting is my assumption.

￼￼`sinfo` probably gets its data directly from slurmctld (last I looked), but you're right this is Slurm's attempt at keeping a historical record of down/drained node events. Not associated with accounting, but we should ask anyone uses that info here.

**Comment by grondo** [comment]
> My understanding is that WCKey is what users charge their jobs against.

I believe WCKey is a separate way for users to classify a job, allowing it to perhaps be charged against a different bank. Time is "normaly" charged against the users default or requested bank.

According this link, WCKey was added nominally to allow users in different banks to charge against the same project.

https://slurm.schedmd.com/wckey.html

**Comment by cmoussa1** [comment]
Here are an example couple rows of a WCKey table:

| creation_time | mod_time   | deleted | is_def | id_wckey | wckey_name | user     |
| ------------- | ----------- | -------- | ------ | -------- | ------------- | ------- |
|    1571112263 | 1571112263 |       0 |      1 |        1 |            | user1 |
|    1571112263 | 1571112263 |       0 |      0 |        2 | *          | user1 |
|    1572045636 | 1572045636 |       0 |      1 |        3 |            | user2 |
|    1572045636 | 1572045636 |       0 |      0 |        4 | *          | user2 |

**Comment by cmoussa1** [comment]
Going to give a short summary of @dongahn and I's meeting yesterday concerning the structure of the accounting database. If I miss anything, @dongahn, feel free to add/correct me:

The **wckey** table allows users to associate their jobs with the same projects, but from a different bank. 

In terms of limits that are set in Slurm, there are:

**Partition limits**: maxnodes, minnodes, walltime. These are all per-job.
 
**QOS limits**: maxTRESPU=node=… This sets a _maxnodes_ per user across all of their jobs, e.g. on the pdebug queues we limit users to no more than ½ of the nodes in the queue. We also set various flags on some QOS to allow them to override partition limits and/or limits from another QOS.
 
**Association limits**: grpTRES=cpu=… This sets a max cpus (nodes) across all of a user’s jobs in a given bank. It’s possible that these could be implemented as per-user limits on a queue. We will also sometimes set MaxJobs=0 on a user as a quick and dirty way to prevent them from submitting new jobs. We also set a maxsubmit limit here as a barrier against users with runaway scripts.

**Comment by cmoussa1** [comment]
Here is a pass at designing our multi-table DB:

**QOS Table**
Description: one can specify a Quality of Service (QOS) for each job submitted to Flux that affects its priority. 

| Field                     | Type                | Null | Key | Default | Extra          |
| - | - | - | - | - | - |
| creation_time             | bigint(20) unsigned | NO   |     | NULL    |                |
| mod_time                  | bigint(20) unsigned | NO   |     | 0       |                |
| deleted                   | tinyint(4)          | YES  |     | 0       |                |
| id                        | int(11)             | NO   | PRI | NULL    | auto_increment |
| name                      | tinytext            | NO   | UNI | NULL    |                |
| description               | text                | YES  |     | NULL    |                |
| flags                     | int(10) unsigned    | YES  |     | 0       |                |
| grace_time                | int(10) unsigned    | YES  |     | NULL    |                |
| max_jobs_pa               | int(11)             | YES  |     | NULL    |                |
| max_tres_pa               | text                | NO   |     | NULL    |                |
| max_tres_pu               | text                | NO   |     | NULL    |                |
| priority                  | int(10) unsigned    | YES  |     | 0       |                |

**TRES Table**
Description: defines and labels the resources that are used by jobs (specific usage of the resources are found in the **job table**). 

| Field         | Type                | Null | Key | Default | Extra          |
| - | - | - | - | - | - | 
| creation_time | bigint(20) unsigned | NO   |     | NULL    |                |
| deleted       | tinyint(4)          | NO   |     | 0       |                |
| id            | int(11)             | NO   | PRI | NULL    | auto_increment |
| type          | tinytext            | NO   | MUL | NULL    |                |
| name          | tinytext            | NO   |     | NULL    |                |

**User Table**
Description: lists users and their roles on the machine (operator, coordinator, or admin)

| Field         | Type                | Null | Key | Default | Extra |
| - | - | - | - | - | - |
| creation_time | bigint(20) unsigned | NO   |     | NULL    |       |
| mod_time      | bigint(20) unsigned | NO   |     | 0       |       |
| deleted       | tinyint(4)          | YES  |     | 0       |       |
| user_name          | tinytext            | NO   | PRI | NULL    |       |
| admin_level   | smallint(6)         | NO   |     | 1       |       |

**<cluster name> Association Table**
Description: lists associations on a cluster (i.e. a combination of a cluster, account, user name, and optional partition name), defines a user's limits when submitting jobs.

Field Name | Type | NULL | Key | Default | Extra | Notes
-- | -- | -- | -- | -- | -- | --
creation_time | bigint (20) unsigned | NO |   | NULL |   |  
mod_time | bigint (20) unsigned | NO |   | 0 |   |  
deleted | tinyint (4) | NO |   | 0 |   |  
id_assoc | int (10) unsigned | NO | PRI | NULL | auto_increment |  
user | tinytext | NO | MUL | NULL |   |  
acct | tinytext | NO | MUL | NULL |   | 
partition         | tinytext            | NO   |     | NULL    |                |
parent_acct       | tinytext            | NO   |     | NULL    |                | 
shares | int (11) | NO |   | 1 |   |  
max_jobs          | int(11)             | YES  |     | NULL    |                |
max_tres_pj       | text                | NO   |     | NULL    |                |
max_wall_pj       | int(11)             | YES  |     | NULL    |                |
grp_tres          | text                | NO   |     | NULL    |                |
grp_wall          | int(11)             | YES  |     | NULL    |                |
qos               | blob                | NO   |     | NULL    |                |

**<cluster name> WCKey Table**
Description: Allows a user in a different bank to charge against the same project.

Field Name | Type | NULL | Key | Default | Extra | Notes
-- | -- | -- | -- | -- | -- | --
creation_time | bigint (20) unsigned | NO |   | NULL |   |  
mod_time | bigint (20) unsigned | NO |   | 0 |   |  
deleted | tinyint (4) | NO |   | 0 |   |  
is_def | tinyint (4) | NO |   | 0 |   |  
id_wckey | int (10) unsigned | NO | PRI | NULL | auto_increment |  
wckey_name | tinytext | NO | MUL | NULL |   |  
user | tinytext | NO |   | NULL |   |  

**<cluster name> Job Table**
Description: Stores an entry for every job that runs on the cluster

Field Name | Type | NULL | Key | Default | Extra | Notes
-- | -- | -- | -- | -- | -- | --
job_db_inx | bigint (20) unsigned | NO |   |   |   |  
mod_time | bigint (20) unsigned | NO |   |   |   |  
deleted | tinyint (4) | NO |   |   |   |  
cpus_req | int (10) unsigned | NO |   |   |   |  
exit_code | int (10) unsigned | NO |   |   |   |  
job_name | tinytext | NO |   |   |   |  
id_assoc | int (10) unsigned | NO |   |   |   |  
id_job | int (10) unsigned | NO |   |   |   |  
id_qos | int (10) unsigned | NO |   |   |   |  
id_wckey | int (10) unsigned | NO |   |   |   |  
id_user | int (10) unsigned | NO |   |   |   |  
id_group | int (10) unsigned | NO |   |   |   |  
nodelist | text | YES |   |   |   |  
nodes_alloc | int (10) unsigned | NO |   |   |   |  
partition | tinytext | NO |   |   |   |  
priority | int (10) unsigned | NO |   |   |   |  
state | int (10) unsigned | NO |   |   |   |  
timelimit | int (10) unsigned | NO |   |   |   |  
time_submit | bigint (20) unsigned | NO |   |   |   |  
time_sched | bigint (20) unsigned | NO |   |   |   |  
time_depend | bigint (20) unsigned | NO |   |   |   |  
time_start | bigint (20) unsigned | NO |   |   |   |  
time_end | bigint (20) unsigned | NO |   |   |   |  
work_dir | text |   |   |   |   |  
tres_alloc | text |   |   |   |   |  
tres_req | text |   |   |   |   |  

**<cluster name> Association Usage (Hour | Day | Month)**
Description: Track an association's usage on a cluster over a specified period of time.

Field Name | Type | NULL | Key | Default | Extra | Notes
-- | -- | -- | -- | -- | -- | --
creation_time | bigint (20) unsigned | NO |   | NULL |   |  
mod_time | bigint (20) unsigned | NO |   | 0 |   |  
deleted | tinyint (4) | NO |   | 0 |   |  
id | int (10) unsigned | NO | PRI | NULL |   |  
id_tres | int (11) | NO | PRI | 1 |   |  
time_start | bigint (20) unsigned | NO | PRI | NULL |   |  
alloc_secs | bigint (20) unsigned | NO |   | 0 |   |
